linera-service/src/chain_listener.rs

```diff
@@ -46,55 +74,135 @@ where
-        let mut tracker = NotificationTracker::default();
-        while let Some(notification) = notification_stream.next().await {
-            if tracker.insert(notification.clone()) {
-                info!("Received new notification: {:?}", notification);
-                if self.config.delay_before_ms > 0 {
-                    tokio::time::sleep(Duration::from_millis(self.config.delay_before_ms)).await;
-                }
-                {
-                    let mut client = self.client.lock().await;
-                    match &notification.reason {
-                        Reason::NewBlock { .. } => {
-                            if let Err(e) = client.update_validators().await {
-                                warn!(
-                                    "Failed to update validators about the local chain after \
-                                    receiving notification {:?} with error: {:?}",
-                                    notification, e
-                                );
-                            }
-                        }
-                        Reason::NewIncomingMessage { .. } => {
-                            if let Err(e) = client.process_inbox().await {
-                                warn!(
-                                    "Failed to process inbox after receiving new message: {:?} \
-                                    with error: {:?}",
-                                    notification, e
-                                );
-                            }
-                        }
+        let mut streams = HashMap::new();
+        self.update_streams(&mut streams, &mut context, &storage)
+            .await?;
+
+        while let (Some((notification, client)), _, _) =
+            future::select_all(streams.values_mut().map(StreamExt::next)).await
+        {
+            if let Reason::NewBlock { hash, .. } = notification.reason {
+                let value = storage.read_value(hash).await?;
+                let executed_block = &value.inner().executed_block();
+                let timestamp = executed_block.block.timestamp;
+                for out_msg in &executed_block.messages {
+                    if let OutgoingMessage {
+                        destination: Destination::Recipient(new_id),
+                        message: Message::System(SystemMessage::OpenChain { public_key, .. }),
+                        ..
+                    } = out_msg
+                    {
+                        let key_pair = context.wallet_state().key_pair_for_pk(public_key);
+                        context.update_wallet_for_new_chain(*new_id, key_pair, timestamp);
                     }
-                    context.update_wallet(client.deref_mut()).await;
                 }
+            }
+            context.update_wallet(&mut *client.lock().await).await;
+            // TODO(#901): Node service should track newly opened chains
+            // self.update_streams(&mut streams, &mut context, &storage)
```

Comment:
Remove commented code (and this TODO?)

Reply:
Yes, it's convenient to keep them, so it's easier to reproduce the issue.